### DISCUSSION TITLE
How do we keep the closure clean with the `observe`?

### DISCUSSION THREAD

**Discussion by y-mimura** [pointfreeco/swift-composable-architecture#2733] [Q&A] [answered]
I'm currently in the process of migrating to 1.7, but faced the challenge of the title.
I had been used a lot of `store.publisher` or `store.scope.ifLet` on UIViewController to observe state changes before 1.7. For instance:

```swift
override func viewDidLoad() {
  setupObservers()
}

private func setupObservers() {
  observeAppearance()
  observeLogin()
  // ...
}

private func observeAppearance() {
  store.publisher.appearance.sink {
    // ....
  }
}

private func observeLogin() {
  store.scope(
    state: \.login,
    action: \.login
  ).ifLet {
    // ...
  }
}

// ...
```

I'm replacing these API with `observe`, however, I would like to know how the code should be split up.

For instance, should one `observe` contain multiple functions?
```swift
override func viewDidLoad() {
  observe { [weak self] in
    guard let self else { return } 
    observeAppearance()
    observeLogin()
    // ...
  }
}

private func observeAppearance() {
  let appearance = store.appearance
  // ....
}

private func observeLogin() {
  if let loginStore = store.scope(state: \.login, action: \.login) {
    // ...
  }
}

// ...
```

Or should we use multiple `observe`s?

```swift
override func viewDidLoad() {
  observeAppearance()
  observeLogin()
}

private func observeAppearance() {
  observe { [weak self] in
    guard let self else { return } 
    let appearance = store.appearance
    // ....
  }
}

private func observeLogin() {
  observe { [weak self] in
    guard let self else { return } 
    if let loginStore = store.scope(state: \.login, action: \.login) {
      // ...
    }
  }
}

// ...
```

**Comment by acosmicflamingo** [comment]
I had this question as well as I was migrating my UIKit codebase to using the new Observability framework. What Brandon had told me is that in general, you should be safe to add everything in a single `observe` closure and that executing something like `myButton.title = store.name` multiple times is safe. If you were to be doing something more computationally intensive like reloading a table view, then you would want to have separate observe closures. This discussion might help you here: https://github.com/pointfreeco/swift-composable-architecture/discussions/2722.

**Reply by y-mimura** [reply]
@acosmicflamingo
Thank you for the reply! Actually, after reading the discussion https://github.com/pointfreeco/swift-composable-architecture/discussions/2722, I thought I should split it up.
In my case, I had been used `store.scope.ifLet` to decide that should execute [present](https://developer.apple.com/documentation/uikit/uiviewcontroller/1621380-present) or not. Therefore, I believe that `observe` should be separated in my case since `present` might be executed multiple times if the `store.appearance` (in my example) was changed.

**Reply by acosmicflamingo** [reply]
Glad I can help! That certainly sounds like what I'd do too ;) Good luck!

**Reply by mbrandonw** [reply, marked as answer]
Hi @y-mimura, while it is completely fine to split out into multiple `observe`s, it is also possible to put it in one and do navigation. Our TicTacToe demo shows how to do this:

https://github.com/pointfreeco/swift-composable-architecture/blob/4c6b617f8868fd65b4090aa1877cdb120d07795f/Examples/TicTacToe/tic-tac-toe/Sources/LoginUIKit/LoginViewController.swift#L102-L123

In one single `observe` block we are configuring various UI components, as well as presenting/dismissing an alert and drill-down. You just need to do some extra `nil` checking to make sure it is time to do the presenting or dismissing.

**Reply by y-mimura** [reply]
@mbrandonw Thank you for showing this demo. That was exactly what I'd like to know. However, one thing I have wondered about in this demo.

The conditions for displaying alerts are as follows:
https://github.com/pointfreeco/swift-composable-architecture/blob/4c6b617f8868fd65b4090aa1877cdb120d07795f/Examples/TicTacToe/tic-tac-toe/Sources/LoginUIKit/LoginViewController.swift#L102-L103

However, I think it should be divided as follows, is my opinion correct?
```swift
  if let store = store.scope(state: \.alert, action: \.alert) {
    if alertController == nil {
    // ...
```

In my understanding, the closure of the `observe` will be executed each time the referred `store` property was changed. Therefore, if the conditions are combined, when other than the `alert` state was changed during the `alertController` is presented, the `alertController` might be dismissed unexpectedly.

**Reply by mbrandonw** [reply]
Hi @y-mimura, you are right it's not quite correct. The two factor navigation just below is the correct way to do this kind of logic:

https://github.com/pointfreeco/swift-composable-architecture/blob/4c6b617f8868fd65b4090aa1877cdb120d07795f/Examples/TicTacToe/tic-tac-toe/Sources/LoginUIKit/LoginViewController.swift#L112-L123

The alert condition should be more like this:

```swift
if let store = store.scope(state: \.alert, action: \.alert),
  alertController == nil
{
  …
} else if store.alert == nil, alertController != nil {
  …
}
```

In particular, the `else` branch should check that `store.alert` is `nil` before dismissing. I'll PR a fix soon.

**Reply by y-mimura** [reply]
@mbrandonw Ah, it's crystal clear now. I appreciate your explanations 👍
